message.go

```diff
@@ -13,11 +13,21 @@ type message struct {
 	decodeSlots     []interface{}
 }
 
+func decodeIntoMessage(dec decoder, m *message) error {
+	for _, s := range m.decodeSlots {
+		if err := decodeMessage(dec, m, s); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 func decodeMessage(dec decoder, m *message, i interface{}) error {
 	err := dec.Decode(i)
-	if err == nil {
-		m.remainingFields--
-	}
+	// TODO need to verify whether codec.Decode pulls from the reader if
+	// the decode fails, or whether the reader stays in the same state
+	// as before the Decode
+	m.remainingFields--
```

Comment:
did you run into something that caused you to do this? if not, i'd prefer leaving it as-is until the TODO is done